### PR TITLE
Add streamer test

### PR DESCRIPTION
**Pull request by rgantonio**
This PR adds the [SNAX streamer](https://github.com/KULeuven-MICAS/snax-streamer) into the test. It contains:

- A clean up of unnecessary tests that are not needed
- A template for the testbench for the streamer top (because configurations can change)
- Updated Makefile for the template generation
- Updated default configuration file for a simple test
- Cocotb basic test for the streamer (tests basic features for now, contentions will be done with the tcdm)
- Updated `conftest.py` to include the `waves` parameter for waveform dumping
- Updated `snax_util.py` to include utilization functions like register reads and writes, clock and waits, and so on.

Notes:
- The`test_basic_streamer.py` also incorporates the generation of the necessary Chisel files and test benches. So it is quite a bit parametrizeable based on the configuration. However, the actual test is tied to the default configuration. The test is supposed to be a basic check if the address generation is correct anyway.
- Basically, it contains the first feature check: address generation given the parameters.

### CODE REVIEW

**Comment by rgantonio** [comment]
Wait forgot something haha

File: tests/cocotb/test_basic_streamer.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+from cocotb_test.simulator import run
+import pytest
+import snax_util
+import os
+import subprocess
+
+# Configurable design time parameters
+NARROW_DATA_WIDTH = 64
+TCDM_DEPTH = 256
+TCDM_REQ_PORTS = 12
```
Comment (xiaoling-yi) [inline tests/cocotb/test_basic_streamer.py:29]: Be careful! This should match with the Streamer `tcmd_req` ports number

File: util/templates/tb_streamer_top.sv.tpl [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+  localparam int unsigned NrBanks        = 32;
+  localparam int unsigned TCDMSize       = NrBanks * TCDMDepth * (NarrowDataWidth/8);
+  localparam int unsigned TCDMAddrWidth  = $clog2(TCDMSize);
+
+  // clock and resets
+  logic clk_i;
+  logic rst_ni;
+
+  // ports from streamer to accelerator
```
Comment (xiaoling-yi) [inline util/templates/tb_streamer_top.sv.tpl:29]: Change this too.

**Review by xiaoling-yi — COMMENTED** [review]
Good job!

File: Makefile [hunk 3]
```diff
@@ -51,15 +56,23 @@ $(STREAM_GEN_OUT_TOP_PATH):
...
-	python3 util/scripts/template_gen.py --cfg_path="${STREAM_GEN_CFG_PATH}" \
+$(STREAM_GEN_OUT_SV_PATH):
+	${PYTHON} util/scripts/template_gen.py --cfg_path="${STREAM_GEN_CFG_PATH}" \
 	--tpl_path="${STREAM_GEN_TPL_SV_PATH}" \
 	--out_path="${STREAM_GEN_OUT_SV_PATH}"
 
+#-----------------------------
+# Generate Streamer Wrapper Testbench
+#-----------------------------
+gen_stream_top:	$(STREAM_GEN_OUT_SCALA_PATH) $(STREAM_GEN_OUT_TOP_PATH) $(STREAM_GEN_OUT_SV_PATH)
```
[the remark this answers is not in the document]
Comment (rgantonio) [reply Makefile:69]: OK wait haha

File: Makefile [hunk 4]
```diff
@@ -51,15 +56,23 @@ $(STREAM_GEN_OUT_TOP_PATH):
 	@echo "Generates output: ${STREAM_GEN_OUT_TOP_PATH}"
 
 #-----------------------------
-# Generate Streamer Wrapper
+# Generate tb_stream_top.sv
 #-----------------------------
-gen_stream_top:	$(STREAM_GEN_OUT_SCALA_PATH) $(STREAM_GEN_OUT_TOP_PATH)
-	python3 util/scripts/template_gen.py --cfg_path="${STREAM_GEN_CFG_PATH}" \
+$(STREAM_GEN_OUT_SV_PATH):
```
[the remark this answers is not in the document]
Comment (rgantonio) [reply Makefile:61]: The SV was meant for the system Verilog file and the TB was meant for the testbench. Maybe I should put RTL instead hmm...

[the remark this answers is not in the document]
Comment (rgantonio) [reply Makefile:61]: OK changed in next commits

File: util/cfg/streamer_cfg.hjson [hunk 5]
```diff
@@ -1,43 +1,72 @@
...
+    //--------------------------------
+    // FIFO parameters
+    // fifoWidth - FIFO bit-width
+    // fifoDepth - FIFO depth or number of elements
+    //--------------------------------
+    // For FIFO internal reader
     fifoReaderParams: {
-      fifoWidth: [256,256,64],
-      fifoDepth: [2, 2, 2],
+      fifoWidth: [256, 256],
```
[the remark this answers is not in the document]
Comment (rgantonio) [reply util/cfg/streamer_cfg.hjson:22]: Fixed!

File: util/cfg/streamer_cfg.hjson [hunk 6]
```diff
@@ -1,43 +1,72 @@
...
+    //--------------------------------
+    // For Readers
     dataReaderParams:{
-      tcdmPortsNum: [4,4,1],
-      spatialBounds: [[8],[8],[2]],
-      unrollingDim: [2,2,2],
-      spatialDim: [1,1,1],
-      elementWidth: [32,32,32],
-      fifoWidth: [64,64,64]
+      tcdmPortsNum: [4, 4],
```
[the remark this answers is not in the document]
Comment (rgantonio) [reply util/cfg/streamer_cfg.hjson:40]: Fixed!

File: tests/cocotb/test_basic_streamer.py (hunk 1 above)
Comment (rgantonio) [reply tests/cocotb/test_basic_streamer.py:27]: I know! :'( I have no idea how to make this automatic... Not unless I read the files configuration file but that's just too automated. This is meant for the default one for now haha

File: util/templates/tb_streamer_top.sv.tpl (hunk 2 above)
Comment (rgantonio) [reply util/templates/tb_streamer_top.sv.tpl:28]: done!

**Review by xiaoling-yi — APPROVED** [review]

**Comment by rgantonio** [comment]
@xiaoling-yi grazie!
